src/main/java/io/knowledgeassets/myskills/server/user/profile/UserProfileService.java

```diff
@@ -0,0 +1,240 @@
+					log.warn(format("The template file \"%s\" cannot be read. The default template file will be used as a fallback option.", path), ex);
+					return DEFAULT_TEMPLATE_RESOURCE.getInputStream();
+				}
+			}
+			else {
+				log.warn(format("The template file \"%s\" does not exist. The default template file will be used as a fallback option.", path));
+				return DEFAULT_TEMPLATE_RESOURCE.getInputStream();
+			}
+		}
+	}
+
+	private User getUserByReferenceId(String referenceId) {
+		return userRepository.findByReferenceId(referenceId).orElseThrow(() -> {
+			final String[] searchParamsMap = {"referenceId", referenceId};
+			return NoSuchResourceException.builder()
+					.model(Model.USER)
+					.searchParamsMap(searchParamsMap)
+					.build();
+		});
+	}
+
+	private void replacePlaceholders(XWPFDocument document, User user) {
+		replaceTablesPlaceholders(document, user);
+		replaceParagraphsPlaceholders(document, user, document.getParagraphs());
+		ofNullable(document.getFooterList()).ifPresent(footers ->
+				footers.forEach(footer -> {
+					replaceTablesPlaceholders(footer, user);
+				})
+		);
+	}
+
+	private void replaceTablesPlaceholders(IBody body, User user) {
+		ofNullable(body.getTables()).ifPresent(tables ->
+				tables.forEach(table ->
+						ofNullable(table.getRows()).ifPresent(rows ->
+								rows.forEach(row ->
+										ofNullable(row.getTableCells()).ifPresent(cells ->
+												cells.forEach((XWPFTableCell cell) ->
+														replaceParagraphsPlaceholders(cell, user, cell.getParagraphs())
+												)
+										)
+								)
+						)
+				)
+		);
+	}
+
+	private void replaceParagraphsPlaceholders(IBody body, User user, List<XWPFParagraph> paragraphs) {
+		ofNullable(paragraphs).ifPresent(paragraphList -> {
+			for (int idx = 0; idx < paragraphs.size(); idx++) {
+				replaceParagraphPlaceholders(paragraphList.get(idx), user, body);
+			}
+		});
+	}
+
+	private void replaceParagraphPlaceholders(XWPFParagraph p, User user, IBody body) {
+		ofNullable(p.getRuns()).ifPresent(runs ->
+				runs.forEach(r -> {
+							replacePlaceholder(r, ReportPlaceholder.DATE.getName(), SDF.format(new Date()));
+							replacePlaceholder(r, ReportPlaceholder.ACADEMIC_DEGREE.getName(), user.getAcademicDegree());
+							replacePlaceholder(r, ReportPlaceholder.POSITION_PROFILE.getName(), user.getPositionProfile());
+							replaceListPlaceholder(r, ReportPlaceholder.SPECIALIZATIONS.getName(), user.getSpecializations(), p, body);
+							replaceListPlaceholder(r, ReportPlaceholder.LANGUAGES.getName(), user.getLanguages(), p, body);
+							replaceListPlaceholder(r, ReportPlaceholder.INDUSTRY_SECTORS.getName(), user.getIndustrySectors(), p, body);
+							replaceListPlaceholder(r, ReportPlaceholder.CERTIFICATES.getName(), user.getCertificates(), p, body);
+							replaceListPlaceholder(r, ReportPlaceholder.SKILLS.getName(), StreamSupport.stream(userSkillRepository.findByUserIdOrderByCurrentLevelDesc(user.getId()).spliterator(), false)
```

Comment:
The method `replaceParagraphPlaceholders(XWPFParagraph p, User user, IBody body)` seems to be quite a low-method method which is called in several other places, e.g. during processing of each table cell.

I suppose that doing this same query for user skills in each of those method calls can become a performance issue. Maybe it helps already to check whether the "skills" placeholder really exists within the current paragraph and query the repository only if this is the case.

Reply:
Good point! You are right. Thanks.

I changed the methods `io.knowledgeassets.myskills.server.user.profile.UserProfileDocumentService#replacePlaceholder` and `io.knowledgeassets.myskills.server.user.profile.UserProfileDocumentService#replaceListPlaceholder` to accept `java.util.function.Supplier`. Now skills are requested only in case there is a `${skills}` placeholder in the text of a processed `org.apache.poi.xwpf.usermodel.XWPFRun`.
Done.

Reply:
I like that `Supplier` approach. Great idea! 😃